### PR TITLE
MESH-1439: add http_header_service_identity to check service identitiy in healthchecks

### PR DESCRIPTION
**Pull request by BeryJu**
MESH-1439, other half of https://github.com/Yelp/nerve-tools/pull/94

### CODE REVIEW

File: hacheck/handlers.py [hunk 1]
```diff
@@ -124,6 +124,14 @@ def get(self, service_name, port, query):
             host = '127.0.0.1'
         port = header_port if header_port else port
 
+        if config.config['http_header_service_identity'] != "":
+            identity_header = config.config["http_header_service_identity"]
+            identity = self.request.headers.get(identity_header, None)
+            if identity and identity != service_name:
+                self.set_status(503)
```
Comment (BeryJu) [inline hacheck/handlers.py:131]: Maybe we should have a flag to make this a warning only?

Comment (BeryJu) [reply hacheck/handlers.py:131]: very good point, I hadn't thought of that

Comment (Buffer0x7cd) [reply hacheck/handlers.py:131]: sure, I think the easiest method is to inject the registration info in environment variable and then let the hacheck check against the environment variable. This way hacheck doesn’t need to ask kubelet about annotation or have to look anywhere elese.

Comment (BeryJu) [reply hacheck/handlers.py:131]: Actually the readiness check gets all the registrations already (see https://github.com/Yelp/paasta/blob/master/paasta_tools/kubernetes_tools.py#L943), so maybe that would be a cleaner way to go

Comment (Buffer0x7cd) [reply hacheck/handlers.py:131]: But how are you going to share the registration info mentioned in readiness probe with hacheck process ? you will still need to look into the manifest for that ?

File: hacheck/config.py [hunk 2]
```diff
@@ -35,3 +37,13 @@ def load_from(path):
                 constructor, default = DEFAULTS[key]
                 config[key] = constructor(value)
     return config
+
+
+def get_service_registrations():
+    """
+    Check if we've been given service registrations for this service and attempt to parse them
+    """
+    registrations = os.getenv("SERVICE_REGISTRATIONS", None)
```
Comment (Rob-Johnson) [inline hacheck/config.py:46]: can we use `MESH_REGISTRATIONS`?

File: hacheck/handlers.py [hunk 3]
```diff
@@ -124,6 +132,19 @@ def get(self, service_name, port, query):
             host = '127.0.0.1'
         port = header_port if header_port else port
 
+        if config.config['http_header_service_identity'] != "":
```
Comment (Rob-Johnson) [inline hacheck/handlers.py:135]: is it worth doing this validation before of the host/port reading above?

File: hacheck/config.py [hunk 4]
```diff
@@ -18,7 +19,8 @@ def max_or_int(some_str_value):
     'mysql_password': (str, None),
     'rlimit_nofile': (max_or_int, None),
     'allow_remote_spool_changes': (bool, False),
-    'http_headers_to_copy': (list, ['Host'])
+    'http_headers_to_copy': (list, ['Host']),
+    'http_header_service_identity': (str, "x-smartstack-expected-service"),
```
Comment (Rob-Johnson) [inline hacheck/config.py:23]: sorry for nitpick: I'm not sure `expected-service` is the right thing here, and I don't want to overload `service` even more. maybe `target-identity`? second choice would probably be `target-namespace` (given that's what we're actually targetting here), but I think identity keeps things independent of the smartstack implementation details

File: hacheck/handlers.py (hunk 3 above)
Comment (BeryJu) [reply hacheck/handlers.py]: we don't actually need the validation at all

File: hacheck/config.py (hunk 2 above)
Comment (BeryJu) [reply hacheck/config.py]: Actually I already merged a PR in paasta that uses `SERVICE_REGISTRATIONS`, let me fix that also

**Review by Rob-Johnson — APPROVED** [review]

**Pull request merged**